scripts/hal_alignment/maf_to_fasta.py

```diff
@@ -0,0 +1,240 @@
+    maf_src_patt = f'^(?P<genome>{genome_patt})[.](?P<seqid>.+)$'
+    return re.compile(maf_src_patt)
+
+
+def main(maf_file: Union[Path, str], output_dir: Union[Path, str],
+         genomes_file: Union[Path, str] = None) -> None:
+    """Convert each block of a MAF alignment to a FASTA file.
+
+    Args:
+        maf_file: Input MAF file with alignment blocks. The src fields of
+            this MAF file should be of the form '<genome>.<seqid>'.
+        output_dir: Output directory under which FASTA files will be created.
+        genomes_file: File listing the genomes in the input MAF file, one per line. This is used
+            to compile a regex that splits MAF src fields of the form '<genome>.<seqid>'
+            into their component parts. If any of the genomes or their sequences in the
+            MAF file contains a dot ('.'), this is required.
+
+    Raises:
+        ValueError: If a MAF src field cannot be parsed.
+
+    """
+
+    if genomes_file is not None:
+        with open(genomes_file) as f:
+            genome_names = [line.rstrip() for line in f]
+        maf_src_regex = compile_maf_src_regex(genome_names)
+
+    out_dir_path = Path(output_dir)
+
+    maf_src_map: Dict[str, Tuple[str, str]] = {}
+    with open(maf_file) as in_f:
+        for idx, msa in enumerate(MafIterator(in_f)):
+            idx_dir_path = out_dir_path / map_uint_to_path(idx)
+            os.makedirs(idx_dir_path, exist_ok=True)
+
+            ga_recs = []
+            fasta_file_path = idx_dir_path / f'{idx}.fa'
+            with open(fasta_file_path, 'w') as out_f:
+                for rec in msa:
+                    maf_src = rec.id
+
+                    try:
+                        genome_name, dnafrag_name = maf_src_map[maf_src]
+                    except KeyError:
+                        try:
+                            genome_name, dnafrag_name = maf_src.split('.')
+                            maf_src_map[maf_src] = (genome_name, dnafrag_name)
+                        except ValueError as exc:
+                            if genomes_file is None:
+                                raise ValueError(
+                                    "MAF src field parse failed due to multiple dot separators"
+                                    " - please set genome names with the '--genomes-file' parameter"
+                                ) from exc
+
+                            match = maf_src_regex.match(maf_src)
+                            try:
+                                genome_name = match['genome']  # type: ignore
+                                dnafrag_name = match['seqid']  # type: ignore
+                            except TypeError as exc:
+                                raise ValueError(
+                                    "MAF src regex failed to parse MAF src field: '{maf_src}'") from exc
+
+                            maf_src_map[maf_src] = (genome_name, dnafrag_name)
```

Comment:
Not a huge change, but I find this way you can avoid one `try-except` (and retains readability):
```
                    try: 
                        genome_name, dnafrag_name = maf_src_map.get(maf_src, maf_src.split('.'))
                    except ValueError as exc:
                        if not genomes_file:
                            raise ValueError(
                                "MAF src field parse failed due to multiple dot separators"
                                " - please set genome names with the '--genomes-file' parameter"
                            ) from exc
                        match = maf_src_regex.match(maf_src)
                        try:
                            genome_name = match['genome']  # type: ignore
                            dnafrag_name = match['seqid']  # type: ignore
                        except TypeError as exc:
                            raise ValueError(
                                "MAF src regex failed to parse MAF src field: '{maf_src}'") from exc
                    if maf_src not in maf_src_map:
                        maf_src_map[maf_src] = (genome_name, dnafrag_name)
```
Great example of EAFP :nerd_face: